### PR TITLE
TBufferedTransport must have underlying transport

### PR DESCRIPTION
**Pull request by RobberPhex**
ref: https://issues.apache.org/jira/browse/THRIFT-4477

### CODE REVIEW

File: lib/php/lib/Factory/TStringFuncFactory.php [hunk 1]
```diff
@@ -21,8 +21,9 @@
 
 namespace Thrift\Factory;
 
-use Thrift\StringFunc\Mbstring;
 use Thrift\StringFunc\Core;
+use Thrift\StringFunc\Mbstring;
```
Comment (jeking3) [inline lib/php/lib/Factory/TStringFuncFactory.php]: These changes seem unnecessary in order to change the line of documentation on line 53 below.

File: lib/php/lib/Transport/TBufferedTransport.php [hunk 2]
```diff
@@ -22,6 +22,7 @@
 
 namespace Thrift\Transport;
 
+use Thrift\Exception\TTransportException;
```
Comment (jeking3) [inline lib/php/lib/Transport/TBufferedTransport.php]: I don't see where TTransportException is used in the changes you made.
Is a "use" clause needed for the @inheritdoc directive to work?

File: lib/php/lib/Transport/TTransport.php [hunk 3]
```diff
@@ -22,6 +22,7 @@
 
 namespace Thrift\Transport;
 
+use Thrift\Exception\TTransportException;
```
Comment (jeking3) [inline lib/php/lib/Transport/TTransport.php]: Why is this change needed?

File: lib/php/lib/Transport/TBufferedTransport.php [hunk 4]
```diff
@@ -78,11 +69,26 @@ public function __construct($transport = null, $rBufSize = 512, $wBufSize = 512)
      */
     protected $rBuf_ = '';
 
+    /**
+     * Constructor. Creates a buffered transport around an underlying transport
+     */
+    public function __construct($transport, $rBufSize = 512, $wBufSize = 512)
```
Comment (jeking3) [inline lib/php/lib/Transport/TBufferedTransport.php]: It would have been more obvious if the pull request changed only this line, removing the "= null".

Comment (RobberPhex) [reply lib/php/lib/Transport/TBufferedTransport.php]: Firstly, remove `=null` at class filed.
And, `$transport=null` should also be removed at function `__construct`.

Sometimes, called `new TBufferedTransport()` will cause $transport_ is null, which cause other funcion doesn't work(`open`, `read`, etc.)

File: lib/php/lib/Factory/TStringFuncFactory.php (hunk 1 above)
Comment (RobberPhex) [reply lib/php/lib/Factory/TStringFuncFactory.php]: I sorted use alphabetically, is it unnecessary? Am I expected that change it back?

File: lib/php/lib/Transport/TBufferedTransport.php (hunk 2 above)
Comment (RobberPhex) [reply lib/php/lib/Transport/TBufferedTransport.php]: some function has doc comment `@throws TTransportException`, No one (and IDE) knows which TTransportException, so I think we should add use stmt.

Or, We always use full classname at doc comment, is it?

File: lib/php/lib/Transport/TTransport.php (hunk 3 above)
Comment (RobberPhex) [reply lib/php/lib/Transport/TTransport.php]: There is doc comment `@throws TTransportException` at Line 45.

No one (and IDE) knows which TTransportException.

File: lib/php/lib/Factory/TStringFuncFactory.php (hunk 1 above)
Comment (jeking3) [reply lib/php/lib/Factory/TStringFuncFactory.php]: In general fewer changes are better but in this case no, I don't care.  I do the same cleanup when I'm in files sometimes. :)

File: lib/php/lib/Transport/TBufferedTransport.php (hunk 2 above)
Comment (jeking3) [reply lib/php/lib/Transport/TBufferedTransport.php]: My question was more for learning, and whether or not the "use" is needed for that to work.  I think I understand it now.

**Comment by RobberPhex** [comment]
CI isn't stable.

It's so headache. :(
